---
title: "Biostat M280 Homework 2"
subtitle: Due Feb 16 @ 11:59PM
output: html_document
author: Lumi Huang
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## Q1

Read [Chapter 7](http://r4ds.had.co.nz/exploratory-data-analysis.html) (Exploratory Data Analysis) of _R for Data Science_ and do exercises 7.3.4, 7.4.1, 7.5.1.1, 7.5.2.1, and 7.5.3.1.

7.3.4
1.Explore the distribution of each of the x, y, and z variables in diamonds. What do you learn? Think about a diamond and how you might decide which dimension is the length, width, and depth.  


```{r}
if (!require("tidyverse")) 
     install.packages("tidyverse", repos='http://cran.us.r-project.org')
if (!require("viridis")) 
     install.packages("viridis", repos='http://cran.us.r-project.org')

library(tidyverse)
library(viridisLite)
library(forcats)
library(nycflights13)
#examining the distribution of x, y, z
ggplot(data = diamonds) + geom_histogram(mapping = aes(x), binwidth = 0.5)
ggplot(data = diamonds) + geom_histogram(mapping = aes(y), binwidth = 0.5) + 
  coord_cartesian(xlim = c(0, 15))
ggplot(data = diamonds) + geom_histogram(mapping = aes(z), binwidth = 0.5) +
  coord_cartesian(xlim = c(0,10))
```
I learnt that there are some outliers in y and z by the evidence of the unusually wide limits on the x-axis. Based on the distribution of x, y, z without outliers, we can say x, y are the length and width of diamonds, and z is the depth of diamonds.


2.Explore the distribution of price. Do you discover anything unusual or surprising? (Hint: Carefully think about the binwidth and make sure you try a wide range of values.)  

```{r}
#small value for binwidth
ggplot(data = diamonds) + 
  geom_histogram(mapping = aes(x = price), binwidth = 10)

#try a wide range of values
ggplot(data = diamonds) + 
  geom_histogram(mapping = aes(x = price), binwidth = 150)

```
The distribution of price is very positively skewed. It looks like there is a lower bound for the price of diamonds based on the distribution of price.


3.How many diamonds are 0.99 carat? How many are 1 carat? What do you think is the cause of the difference?
```{r}
carat_0.99 <- diamonds %>%
  filter(carat == 0.99)
count(carat_0.99)

carat_1 <- diamonds %>%
  filter(carat == 1)
count(carat_1)

mean(carat_0.99$price)
mean(carat_1$price)


```
There are 23 diamonds that are 0.99 carat, and 1558 diamonds that are 1 carat. The price differences between 0.99 carat and 1 carat diamonds could be the cause of this difference. Because 1 carat diamond is about $800 more expensive than 0.99 carat diamond, we expect to see more 1 carat diamonds than 0.99 carat diamonds even if the other figures are about the same.


4.Compare and contrast coord_cartesian() vs xlim() or ylim() when zooming in on a histogram. What happens if you leave binwidth unset? What happens if you try and zoom so only half a bar shows?
```{r}
#coor_cartesian
ggplot(data = diamonds) + 
  geom_histogram(mapping = aes(x = price), binwidth = 100) +
  coord_cartesian(xlim = c(0, 10000))

#xlim()
ggplot(data = diamonds) +
  geom_histogram(mapping = aes(x = price), binwidth = 100) + xlim(0, 10000)



#leave binwidth unset
ggplot(data = diamonds) + 
  geom_histogram(mapping = aes(x = price)) +
  coord_cartesian(xlim = c(0, 5000))

#zoom so only half a bar shows
ggplot(data = diamonds) + 
  geom_histogram(mapping = aes(x = price), binwidth = 100) +
  coord_cartesian(xlim = c(0, 50))
```

Coor_cartesian and xlim() generate the same shape when zooming in on a histogram. The difference lies on how they process the zoom-in command. Coor_cartesian simply zoom in the data in the given limits (the histogram is not affected) while xlim() removes the data that fall outside of the boundary first, then calculate the histogram.
When we leave binwidth unset, r studio uses bins = 30 as default setting.
When the boundary contains only half a bar, rstudio does not show anything on the historgram.


7.4.1
1.What happens to missing values in a histogram? What happens to missing values in a bar chart? Why is there a difference?
```{r}
#histogram
diamonds %>%
  mutate(x = ifelse(x < 3 | x > 15, NA, x)) %>%
  ggplot(data = diamonds, mapping = aes(x)) + 
  geom_histogram(binwidth = 0.5)


#bar
diamonds %>%
  mutate(cut = if_else(runif(n()) < 0.1, NA_character_,
                       as.character((cut)))) %>%
  ggplot(diamonds, mapping = aes(x = cut)) +
  geom_bar()

```

Missing values in a histogram are removed when calculating the histogram. We can see that from the warning message "Removed 8 rows containing non-finite values (stat_bin)."
Missing values in a bar chart are treated as a new categories because the x-axis variable should be a discrete value for bar chart.


2.What does na.rm = TRUE do in mean() and sum()?
```{r}
mean(c(1, 3, 6, NA, 7), na.rm = TRUE)
sum(c(5, 8, 2, NA, NA, 23), na.rm = TRUE)
```
na.rm = TRUE removes the NA values before calculating mean and sum


7.5.1.1
1.Use what you’ve learned to improve the visualisation of the departure times of cancelled vs. non-cancelled flights.
```{r}
#change count to density
nycflights13::flights %>% 
  mutate(
    cancelled = is.na(dep_time),
    sched_hour = sched_dep_time %/% 100,
    sched_min = sched_dep_time %% 100,
    sched_dep_time = sched_hour + sched_min / 60
  ) %>% 
  ggplot(mapping = aes(x = sched_dep_time, y = ..density..)) + 
  geom_freqpoly(mapping = aes(colour = cancelled), binwidth = 1/4)

#boxplot
nycflights13::flights %>% 
  mutate(
    cancelled = is.na(dep_time),
    sched_hour = sched_dep_time %/% 100,
    sched_min = sched_dep_time %% 100,
    sched_dep_time = sched_hour + sched_min / 60
  ) %>% 
  ggplot() +
  geom_boxplot(mapping = aes(x = cancelled, y = sched_dep_time))

```
We can see cancelled flights have a higher scheduled departure time than non-cancelled flights.


2.What variable in the diamonds dataset is most important for predicting the price of a diamond? How is that variable correlated with cut? Why does the combination of those two relationships lead to lower quality diamonds being more expensive?
```{r}
library(quantreg)
diamonds
#carat: y
diamonds %>%
  ggplot()+
  geom_quantile(mapping = aes(x = carat, y = price))

#cut: n
diamonds %>%
  ggplot() +
  geom_boxplot(mapping = aes(x = reorder(cut, price, FUN = median), y = price))

#color: n
diamonds %>%
  ggplot() +
  geom_boxplot(mapping = aes(x = reorder(color, price, FUN = median), y = price))

#clarity:n
diamonds %>%
  ggplot() +
  geom_boxplot(mapping = aes(x = reorder(clarity, price, FUN = median), y = price))

#depth: n
diamonds %>%
  ggplot()+
  geom_quantile(mapping = aes(x = depth, y = price))
#table: y
diamonds %>%
  ggplot()+
  geom_quantile(mapping = aes(x = table, y = price))
#x: n
diamonds %>%
  ggplot()+
  geom_quantile(mapping = aes(x = x, y = price))
#y: y
diamonds %>%
  ggplot()+
  geom_quantile(mapping = aes(x = y, y = price))
#z: y
diamonds %>%
  ggplot()+
  geom_quantile(mapping = aes(x = z, y = price))

#cut and carat:
diamonds %>%
  ggplot() +
  geom_boxplot(mapping = aes(x = reorder(cut, carat, FUN = median), y = carat))

```
Carat is the most important for predicting the price of a diamond. Carat and the price of the diamond have a positive relationship. When there is an increase in carat, the price of the diamond increases as well. We examined the relationship between cut and carat. We can see that the fair diamonds have higher carat in average, and higher carat is related to higher diamond price. The combination of these two relationship lead to lower quality diamonds being more expensive.



3.Install the ggstance package, and create a horizontal boxplot. How does this compare to using coord_flip()?
```{r}
#coord_flip()
diamonds %>%
  ggplot() +
  geom_boxplot(mapping = aes(x = reorder(cut, carat, FUN = mean), y = carat)) +
  coord_flip()

#ggstance
library(ggstance)
ggplot(data = diamonds) +
  geom_boxploth(mapping = aes(y = reorder(cut, carat, FUN = mean), x = carat))
  
```
The outputs are exactly the same. The only difference is in the way we labeled x and y axis.


4.One problem with boxplots is that they were developed in an era of much smaller datasets and tend to display a prohibitively large number of “outlying values”. One approach to remedy this problem is the letter value plot. Install the lvplot package, and try using geom_lv() to display the distribution of price vs cut. What do you learn? How do you interpret the plots?
```{r}
if (!require("lvplot")) 
     install.packages("lvplot", repos='http://cran.us.r-project.org')

library(lvplot)
diamonds %>%
  ggplot() + 
  geom_lv(mapping = aes(x = cut, y = price))
```
Compared to boxplot, letter value plot gives more information about the distribution.

5.Compare and contrast geom_violin() with a facetted geom_histogram(), or a coloured geom_freqpoly(). What are the pros and cons of each method?
```{r}
diamonds %>%
  ggplot() + 
  geom_violin(mapping = aes(x = cut, y = price))

diamonds %>%
  ggplot() + 
  geom_histogram(mapping = aes(x = price)) + 
  facet_wrap(~cut)
```
Violin plot shows multimodal data, and is better in discribing a distribution. However, in violin plot, we cannot find the number of observations count. While in histogram, we can see the number of observations, but the bin size may change the shape of the distribution.

6.If you have a small dataset, it’s sometimes useful to use geom_jitter() to see the relationship between a continuous and categorical variable. The ggbeeswarm package provides a number of methods similar to geom_jitter(). List them and briefly describe what each one does.
```{r}
if (!require("ggbeeswarm")) 
     install.packages("ggbeeswarm", repos='http://cran.us.r-project.org')
library(ggbeeswarm)

diamonds %>%
  ggplot() + 
  geom_quasirandom(mapping = aes(x = cut, y = price))

```
It is similar to violin on how it expressed the density with width.

7.5.2.1
1.How could you rescale the count dataset above to more clearly show the distribution of cut within colour, or colour within cut?
```{r}
diamonds %>% 
  count(color, cut) %>% 
  group_by(color) %>%
  mutate(percentile = n / sum(n)) %>%
  ggplot(mapping = aes(x = color, y = cut)) +
  geom_tile(mapping = aes(fill = percentile))
```
Rescale the count dataset to percentile shows how the data is distributed. For instance, we can see diamonds with ideal cut and color D have a higher percentile.

2.Use geom_tile() together with dplyr to explore how average flight delays vary by destination and month of year. What makes the plot difficult to read? How could you improve it?
```{r}

```

3.Why is it slightly better to use aes(x = color, y = cut) rather than aes(x = cut, y = color) in the example above?
```{r}

```


#7.5.3.1
1.Instead of summarising the conditional distribution with a boxplot, you could use a frequency polygon. What do you need to consider when using cut_width() vs cut_number()? How does that impact a visualisation of the 2d distribution of carat and price?
```{r}

```


2.Visualise the distribution of carat, partitioned by price.
```{r}

```


3.How does the price distribution of very large diamonds compare to small diamonds. Is it as you expect, or does it surprise you?
```{r}

```


4.Combine two of the techniques you’ve learned to visualise the combined distribution of cut, carat, and price.
```{r}

```


5.Two dimensional plots reveal outliers that are not visible in one dimensional plots. For example, some points in the plot below have an unusual combination of x and y values, which makes the points outliers even though their x and y values appear normal when examined separately.
```{r}

```


## Q3
Redo HW1 Q2 using tidyverse.

1. How many persons are in the data set (statisticians call this `n`)? How many SNPs are in the data set (statisticians call this `p`)?
```{r}
setwd(".")
mergegeno_bim <- read.table("/home/m280-data/hw1/merge-geno.bim",
                            col.names = c("Chromosome", "SNPID", 
                                          "GeneticDistance", 
                                          "BasePairPosition", "Allele1", 
                                          "Allele2"))
mergegeno_fam <- read.table("/home/m280-data/hw1/merge-geno.fam",
                            col.names = c("FamilyID", "PersonID", "FatherID",
                                          "MotherID", "Sex", "AffectionStatus"))

#numbers of persons in the data set
mergegeno_fam %>%
  count()
#numbers of SNPs in the data set
mergegeno_bim %>%
  count()

```


2. Which chromosomes does this data set contain? How many SNPs are in each chromosome?
```{r}
#chromsomes
mergegeno_bim %>%
  count(Chromosome)
```



3. MAP4 (microtubule-associated protein 4) is a gene on chromosome 3 spanning positions 47,892,180 bp -- 48,130,769 bp. How many SNPs are located within MAP4 gene?
```{r}
mergegeno_bim %>%
  filter(Chromosome == 3 & 
           BasePairPosition >= 47892180 & BasePairPosition <= 48130769) %>%
  count(Chromosome)
```


4. Statistical geneticists often have to reformat a data set to feed into various analysis programs. For example, to use the Mendel software <http://www.genetics.ucla.edu/software/mendel>, we have to reformat the data set to be read by Mendel.

      - Mendel's SNP definition file is similar to the plink `bim` file but has format  
      `SNP ID`, `Chromosome`, `Base Pair Position`  
      with each field separated by a comma. Write a Linux shell command to convert `merge-geno.bim` to Mendel SNP definition file.

```{r}
mendo_bim <- mergegeno_bim %>%
  select(one_of(c("SNPID", "Chromosome", "BasePairPosition"))) %>%
  unite(SNPID, Chromosome, BasePairPosition, col = "", sep = ",")

print("2.40 = FILE FORMAT VERSION NUMBER.")
print("8348674  = NUMBER OF SNPS LISTED HERE.")
head(mendo_bim)
```

    
    - Mendel's pedigree file is similar to the plink `fam` file but has format  
    `Family ID`, `Person ID`, `Father ID`, `Mother ID`, `Sex` coded as M or F, `Twin Status`  
    with each field separated by a comma. Write a Linux shell command to convert `merge-geno.fam` to Mendel pedigree file. Since twin status is not available in plink format, we put nothing for that field. Also Mendel limits Person ID to have length less than or equal to 8 characters, so we have to strip the string `T2DG` from the IDs. 
```{r}
#Family ID, Person ID, Father ID, Mother ID, Sex coded as M or F, Twin Status
mendo_fam <- mergegeno_fam %>%
  mutate(Sex = if_else(Sex == 1, "M", "F", missing = NULL)) %>%
  select(one_of(c("FamilyID", "PersonID", "FatherID", "MotherID", "Sex"))) %>%
  mutate(TwinStatus = "") %>%
  unite(col = "", sep = ",")
head(mendo_fam)
```





